packages/core/src/standard.ts

```diff
@@ -0,0 +1,112 @@
+/** The Standard Schema interface. */
+export interface StandardSchemaV1<Input = unknown, Output = Input> {
+  /** The Standard Schema properties. */
+  readonly "~standard": StandardSchemaV1.Props<Input, Output>;
+}
+
+export declare namespace StandardSchemaV1 {
+  /** The Standard Schema properties interface. */
+  export interface Props<Input = unknown, Output = Input> {
+    /** The version number of the standard. */
+    readonly version: 1;
+    /** The vendor name of the schema library. */
+    readonly vendor: string;
+    /** Validates unknown input values. */
+    readonly validate: (
+      value: unknown,
+    ) => Result<Output> | Promise<Result<Output>>;
+    /** Inferred types associated with the schema. */
+    readonly types?: Types<Input, Output> | undefined;
+  }
+
+  /** The result interface of the validate function. */
+  export type Result<Output> = SuccessResult<Output> | FailureResult;
+
+  /** The result interface if validation succeeds. */
+  export interface SuccessResult<Output> {
+    /** The typed output value. */
+    readonly value: Output;
+    /** The non-existent issues. */
+    readonly issues?: undefined;
+  }
+
+  /** The result interface if validation fails. */
+  export interface FailureResult {
+    /** The issues of failed validation. */
+    readonly issues: ReadonlyArray<Issue>;
+  }
+
+  /** The issue interface of the failure output. */
+  export interface Issue {
+    /** The error message of the issue. */
+    readonly message: string;
+    /** The path of the issue, if any. */
+    readonly path?: ReadonlyArray<PropertyKey | PathSegment> | undefined;
+  }
+
+  /** The path segment interface of the issue. */
+  export interface PathSegment {
+    /** The key representing a path segment. */
+    readonly key: PropertyKey;
+  }
+
+  /** The Standard Schema types interface. */
+  export interface Types<Input = unknown, Output = Input> {
+    /** The input type of the schema. */
+    readonly input: Input;
+    /** The output type of the schema. */
+    readonly output: Output;
+  }
+
+  /** Infers the input type of a Standard Schema. */
+  export type InferInput<Schema extends StandardSchemaV1> = NonNullable<
+    Schema["~standard"]["types"]
+  >["input"];
+
+  /** Infers the output type of a Standard Schema. */
+  export type InferOutput<Schema extends StandardSchemaV1> = NonNullable<
+    Schema["~standard"]["types"]
+  >["output"];
+}
+
+export type StandardSchemaDictionary = Record<string, StandardSchemaV1>;
+export namespace StandardSchemaDictionary {
+  export type InferInput<T extends StandardSchemaDictionary> = {
+    [K in keyof T]: StandardSchemaV1.InferInput<T[K]>;
+  };
+  export type InferOutput<T extends StandardSchemaDictionary> = {
+    [K in keyof T]: StandardSchemaV1.InferOutput<T[K]>;
+  };
+}
+
+export function parseWithDictionary<TDict extends StandardSchemaDictionary>(
```

Comment:
would be nice if we'd accept object validators so we wouldn't need this, but it was hell getting good types with ZodObject compared to Record<...> when I initially made this

Reply:
the alternative would be requiring the user to provide the object schema builder as an option, but that seems a little annoying from a user perspective (and I haven't test to make sure it's possible)
```ts
createEnv({
  ...opts,
  buildObjectSchema: z.object
})
```

Reply:
we need the raw shape to properly check if an accessed variable is shared

Reply:
i think it should be possible to add (although not sure if StandardSchema has a `shape` type like Zod has. it was just a pain and i think a Record<string, Validator> has worked fine